server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -160,4 +384,478 @@ private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request)
+                    }
+
+                    @Override
+                    public Executor executor() {
+                        return EsExecutors.DIRECT_EXECUTOR_SERVICE;
+                    }
+
+                    @Override
+                    public void handleResponse(NodeQueryResponse response) {
+                        if (results instanceof QueryPhaseResultConsumer queryPhaseResultConsumer) {
+                            queryPhaseResultConsumer.addPartialResult(response.topDocsStats, response.mergeResult);
+                        }
+                        for (int i = 0; i < response.results.length; i++) {
+                            var s = request.shards.get(i);
+                            int shardIdx = s.shardIndex;
+                            final SearchShardTarget target = new SearchShardTarget(routing.nodeId(), s.shardId, routing.clusterAlias());
+                            switch (response.results[i]) {
+                                case Exception e -> onShardFailure(shardIdx, target, shardIterators[shardIdx], e);
+                                case SearchPhaseResult q -> {
+                                    q.setShardIndex(shardIdx);
+                                    q.setSearchShardTarget(target);
+                                    onShardResult(q);
+                                }
+                                case null, default -> {
+                                    assert false : "impossible [" + response.results[i] + "]";
+                                }
+                            }
+                        }
+                    }
+
+                    @Override
+                    public void handleException(TransportException e) {
+                        Exception cause = (Exception) ExceptionsHelper.unwrapCause(e);
+                        if (e instanceof SendRequestTransportException || cause instanceof TaskCancelledException) {
+                            // two possible special cases here where we do not want to fail the phase:
+                            // failure to send out the request -> handle things the same way a shard would fail with unbatched execution
+                            // as this could be a transient failure and partial results we may have are still valid
+                            // cancellation of the whole batched request on the remote -> maybe we timed out or so, partial results may
+                            // still be valid
+                            for (ShardToQuery shard : request.shards) {
+                                final int shardIndex = shard.shardIndex;
+                                onShardFailure(
+                                    shardIndex,
+                                    new SearchShardTarget(routing.nodeId(), shard.shardId, routing.clusterAlias()),
+                                    shardIterators[shardIndex],
+                                    e
+                                );
```

Comment:
is there a chance that we can call onNodeQueryFailure here too? Isn't it the same scenario that we already handle when calling getConnection above?

Reply:
Oh right thanks! Sorry, this went on for so long, I missed that duplication, it's the same code now indeed :)